File: CPP/Solutions/2015/1/Year2015Day01.cpp
#include <utility>

#include "Year2015Day01.h"

auto Year2015Day01::parse() {
    return this->input[0];
}

std::string Year2015Day01::part1() {
    auto input = this->parse();
    int floor = 0;
    for (auto i: input) {
        floor += i == '(' ? 1 : -1;
    }
    return std::to_string(floor);
}

std::string Year2015Day01::part2() {
    auto input = this->parse();
    int floor = 0;
    for (int i = 0; i < input.length(); ++i) {
        floor += input[i] == '(' ? 1 : -1;
        if (floor == -1) {
            return std::to_string(i + 1);
        }
    }
    throw std::exception();
}

Year2015Day01::Year2015Day01(std::vector<std::string> lines) {
    this->input = std::move(lines);
}

File: CSharp/Solutions/2015/1/Year2015Day01.cs
using System;
using System.Collections.Generic;
using System.Linq;
using AdventOfCode.Common;

namespace AdventOfCode.Solutions._2015._1
{
	// ReSharper disable once UnusedMember.Global
	public class Year2015Day01 : ISolution
	{
		public string Part1(IEnumerable<string> input)
		{
 			var directions = input.First();
			return directions.Sum(r => r == '(' ? 1 : -1).ToString();
		}

		public string Part2(IEnumerable<string> input)
		{
			var directions = input.First();
			var floor = 0;
			for (var i = 0; i < directions.Length; i++)
			{
				floor += directions[i] == '(' ? 1 : -1;
				if (floor == -1)
					return (i + 1).ToString();
			}
			throw new Exception("Answer not found");
		}
	}
}